The job **`x86_64-gnu-miri`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27634357081/job/81717608108) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/81717608108) [(plain)](https://github.com/rust-lang/rust/commit/eb8f6174a6acfb1c69ddd0ef44c787cb541583af/checks/81717608108/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/pass/shims/x86/rounding-error.rs ... ok
tests/pass/shims/x86/intrinsics-x86-gfni.rs ... ok

FAILED TEST: tests/pass/c-variadic.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-XfFTqD" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Zmiri-force-intrinsic-fallback" "--cfg" "force_intrinsic_fallback" "-O" "-Zmir-opt-level=4" "-Cdebug-assertions=yes" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/tests/pass" "tests/pass/c-variadic.rs" "--edition" "2021"

error: test got exit status: 1, but expected 0
 = note: compilation failed, but was expected to succeed

full stderr:
warning: Miri does not support optimizations: the opt-level is ignored. The only effect of selecting a Cargo profile that enables optimizations (such as --release) is to apply its remaining settings, such as whether debug assertions and overflow checks are enabled.

warning: You have explicitly enabled MIR optimizations, overriding Miri's default which is to completely disable them. Any optimizations may hide UB that Miri would otherwise detect, and it is not necessarily possible to predict what kind of UB will be missed. If you are enabling optimizations to make Miri run faster, we advise using cfg(miri) to shrink your workload instead. The performance benefit of enabling MIR optimizations is usually marginal at best.

error: Undefined Behavior: using alloc1225 as variable argument list pointer but it does not point to a variable argument list
##[error]  --> tests/pass/c-variadic.rs:122:26
   |
LL |             assert_eq!(b.next_arg::<i32>(), 1i32);
   |                          ^^^^^^^^^^^^^^^^^ Undefined Behavior occurred here
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: stack backtrace:
           0: clone_and_advance::variadic
---
Location:
   /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ui_test-0.30.7/src/lib.rs:365

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
   1: <color_eyre[da8e03757f078a96]::config::EyreHook>::into_eyre_hook::{closure#0}<unknown>
      at <unknown source file>:<unknown line>
   2: eyre[279bccad8c2d47c4]::capture_handler<unknown>
      at <unknown source file>:<unknown line>
   3: eyre[279bccad8c2d47c4]::private::format_err<unknown>
      at <unknown source file>:<unknown line>
   4: ui_test[c0ac5c7b4306822f]::run_tests_generic::<ui_test[c0ac5c7b4306822f]::default_file_filter, ui[301ee98d92def767]::run_tests::{closure#1}, alloc[9caed2ba142c949c]::boxed::Box<dyn ui_test[c0ac5c7b4306822f]::status_emitter::StatusEmitter>><unknown>
      at <unknown source file>:<unknown line>
   5: ui[301ee98d92def767]::ui<unknown>
      at <unknown source file>:<unknown line>
   6: ui[301ee98d92def767]::main<unknown>
      at <unknown source file>:<unknown line>
   7: std[1a86cfa8aa7d08ea]::sys::backtrace::__rust_begin_short_backtrace::<fn() -> core[14691da65407fa5c]::result::Result<(), eyre[279bccad8c2d47c4]::Report>, core[14691da65407fa5c]::result::Result<(), eyre[279bccad8c2d47c4]::Report>><unknown>
      at <unknown source file>:<unknown line>
   8: std[1a86cfa8aa7d08ea]::rt::lang_start::<core[14691da65407fa5c]::result::Result<(), eyre[279bccad8c2d47c4]::Report>>::{closure#0}<unknown>
      at <unknown source file>:<unknown line>
   9: std[1a86cfa8aa7d08ea]::rt::lang_start_internal<unknown>
      at <unknown source file>:<unknown line>
  10: main<unknown>
      at <unknown source file>:<unknown line>
  11: __libc_start_main<unknown>
      at <unknown source file>:<unknown line>
  12: _start<unknown>
      at <unknown source file>:<unknown line>

Run with COLORBT_SHOW_HIDDEN=1 environment variable to disable frame filtering.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/ui-2814ac7bbd734e86 tests/pass tests/panic` (exit status: 1)
Bootstrap failed while executing `test --stage 2 src/tools/miri -- tests/pass tests/panic`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo test --target x86_64-unknown-linux-gnu -Zbinary-dep-depinfo -j 4 -Zroot-dir=/checkout --locked --color=always --profile=release --manifest-path /checkout/src/tools/miri/Cargo.toml -- tests/pass tests/panic [workdir=/checkout]` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:191:21
Executed at: src/bootstrap/src/core/build_steps/test.rs:769:19

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: <bootstrap::core::build_steps::test::Miri as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:769:19
   5: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Miri>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1596:36
   6: <bootstrap::core::build_steps::test::Miri as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:695:21
   7: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   8: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
   9: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1139:9
  10: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1118:14
  11: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  12: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  13: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:250:5
  14: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/sys/backtrace.rs:166:18
  15: std::rt::lang_start::<()>::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:206:18
  16: <&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:287:21
  17: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
  18: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:544:19
  19: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panic.rs:359:14
  20: std::rt::lang_start_internal::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:175:24
  21: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
---
  28: __libc_start_main
  29: _start


Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:24
  local time: Tue Jun 16 18:00:32 UTC 2026
  network time: Tue, 16 Jun 2026 18:00:32 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

--- a/compiler/rustc_const_eval/src/interpret/intrinsics.rs
+++ b/compiler/rustc_const_eval/src/interpret/intrinsics.rs
@@ -725,13 +725,19 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
 
             sym::va_copy => {
                 let va_list = self.deref_pointer(&args[0])?;
+
                 let key_mplace = self.va_list_key_field(&va_list)?;
                 let key = self.read_pointer(&key_mplace)?;
 
                 let varargs = self.get_ptr_va_list(key)?;
                 let copy_key = self.va_list_ptr(varargs.clone());
 
-                let copy_key_mplace = self.va_list_key_field(dest)?;
+                // Zero the destination VaList, so it is fully initialized.
+                let dest = self.force_allocation(dest)?;
+                let zeros = std::iter::repeat_n(0u8, dest.layout.size.bytes_usize());
+                self.write_bytes_ptr(dest.ptr(), zeros)?;
+
+                let copy_key_mplace = self.va_list_key_field(&dest)?;
                 self.write_pointer(copy_key, &copy_key_mplace)?;
             }
 

--- a/library/core/src/intrinsics/mod.rs
+++ b/library/core/src/intrinsics/mod.rs
@@ -3641,6 +3641,10 @@ pub const unsafe fn va_arg<T: VaArgSafe>(ap: &mut VaList<'_>) -> T;
 #[rustc_intrinsic]
 #[rustc_nounwind]
 pub const fn va_copy<'f>(src: &VaList<'f>) -> VaList<'f> {
+    // This fallback body exploits the fact that our codegen backends all just use
+    // a plain memcpy to duplicate VaList. This assumption is wrong for Miri.
+    assert!(!cfg!(miri), "fallback body is incorrect under Miri");
+
     src.duplicate()
 }
 

--- a/src/tools/miri/tests/pass/c-variadic.rs
+++ b/src/tools/miri/tests/pass/c-variadic.rs
@@ -1,3 +1,4 @@
+//@run-native
 #![feature(c_variadic)]
 
 use std::ffi::{CStr, VaList, c_char, c_double, c_int, c_long};
@@ -105,6 +106,63 @@ fn various_types() {
     }
 }
 
+fn clone() {
+    if cfg!(force_intrinsic_fallback) {
+        // Skip this test when we use the fallback bodies. The fallback body does
+        // not hook into the Miri allocation bookkeeping for variable argument lists
+        // and would would falsely report UB.
+        return;
+    }
+
+    unsafe extern "C" fn clone_the_va_list(args: ...) {
+        // The implicit `drop` will catch a `VaList` that isn't properly initialized.
+        let _ = args.clone();
+    }
+
+    unsafe { clone_the_va_list(1i32, 2i32) }
+}
+
+fn clone_and_advance() {
+    if cfg!(force_intrinsic_fallback) {
+        // Skip this test when we use the fallback bodies. The fallback body does
+        // not hook into the Miri allocation bookkeeping for variable argument lists
+        // and would would falsely report UB.
+        return;
+    }
+
+    unsafe extern "C" fn variadic(mut a: ...) {
+        unsafe {
+            let mut b = a.clone();
+            assert_eq!(a.next_arg::<i32>(), 1i32);
+            assert_eq!(b.next_arg::<i32>(), 1i32);
+
+            assert_eq!(a.next_arg::<i32>(), 2i32);
+
+            let mut c = a.clone();
+            let mut d = b.clone();
+
+            assert_eq!(a.next_arg::<i32>(), 3i32);
+            assert_eq!(b.next_arg::<i32>(), 2i32);
+            assert_eq!(c.next_arg::<i32>(), 3i32);
+            assert_eq!(d.next_arg::<i32>(), 2i32);
+
+            assert_eq!(c.next_arg::<i32>(), 4i32);
+            assert_eq!(a.next_arg::<i32>(), 4i32);
+
+            drop(c);
+            drop(a);
+
+            assert_eq!(d.next_arg::<i32>(), 3i32);
+            assert_eq!(d.next_arg::<i32>(), 4i32);
+
+            assert_eq!(b.next_arg::<i32>(), 3i32);
+            assert_eq!(b.next_arg::<i32>(), 4i32);
+        }
+    }
+
+    unsafe { variadic(1i32, 2i32, 3i32, 4i32) }
+}
+
 fn main() {
     ignores_arguments();
     echo();
@@ -112,4 +170,6 @@ fn main() {
     forward_by_ref();
     nested();
     various_types();
+    clone();
+    clone_and_advance();
 }

